Add support for Network Information API

tollmanz opened the pull request:
The Network Information API is a revision of the `navigator.connection` interface that attempts to provide a standard, cross-browser interface to access information about network conditions. This commit updates the mobile plugin to source network information from the newer source where
available.

This plugin works by first trying to find connection information from various places depending on the browser. It then tries to map data from the identified interface to variables that have historically been used by Boomerang. Effectively, this commit makes Boomerang aware of the new locations of information from the Network Information API.

Credit for this work goes to @bluesmoon, with assistance from @tollmanz to verify and write tests.

tollmanz commented on plugins/mobile.js:
```diff
@@ -13,13 +27,28 @@ Plugin to capture navigator.connection.type on browsers that support it
+		var k;
+
+		for (k in param_map) {
+			if (k in connection) {
```
@bluesmoon I changed this from `connection.navigator.hasOwnProperty(k)`. That _always_ returned `false` in my testing regardless of whether that prop existed or not. I tried in console and got the same result. Do you see any issue with using `in` instead?

bluesmoon replied:
`in` returns true even for inherited properties.  Also, should this be `connection.navigator`?  If not, then that was the bug.

tollmanz replied:
Sorry...that was a typo in the comment only.

tollmanz commented on tests/page-templates/19-mobile/02-mobile-connection-disabled.js:
```diff
@@ -0,0 +1,23 @@
+			}
+		}
+
+		assert.isTrue(actual);
```
This test fails. I left it here because I would like to still be able to do it. How can you disable the mobile plugin for testing? I noticed that `BOOMR.plugins` does not have a `mobile` property. The plugin is available and working, but `BOOMR` is not aware of it and it does not allow me to disable it like I normally would.

tollmanz replied:
This no longer fails! With the plugin refactor, it can be disabled and the test does what it should.

tollmanz commented on plugins/mobile.js:
```diff
@@ -13,13 +27,28 @@ Plugin to capture navigator.connection.type on browsers that support it
+
+	// If connection information changes, we collect the latest values
+	if (connection.addEventListener) {
+		connection.addEventListener("change", function() {
```
Any thoughts on how this could be tested?

bluesmoon replied:
I think you can call `dispatchEvent` to simulate an event firing.

tollmanz commented:
A slight change in behavior to note about this changeset: in the previous version, if `navigator.connection` was set, you would likely get `mob.ct`, `mob.bw`, and `mob.mt` with empty values in the beacon. In this version, those values are pruned because data is only attached if it is available. This is technically a breaking change, but one with very few consequences IMO.

tollmanz commented:
> This will be a problem if the plugin is loaded before boomerang. I'll make some changes and do a three way merge.

Yes...so this plugin doesn't seem to follow the correct interface for a Boomerang plugin. I left it alone for now thinking that it could be addressed later. If you are ok with tackling this now, I'd be happy to turn this into a proper plugin. I'd like that opportunity to better understand the plugin system.

bluesmoon commented:
Yeah, go ahead.  I'd started to see if there were a way to do this that still let us set the values when the plugin were loaded rather than when boomerang was initialised, but it comes out to the same, and there's probably not much difference in terms of time.  You'll need an `init` and `is_complete` method at the minimum I think.

tollmanz commented:
The plugin is now rewritten as an actual Boomerang plugin. Instead of using the `BOOMR.subscribe` mechanism to set the vars, I just set it on init. These vars should only be set once in the majority of cases. I've kept the same listener in place should the network info change.

Note that I also added handling of the `saveData` property. @bluesmoon This value is `true` or `false` according to the spec. Do you want that data sent as `(string) true/false` or should I alter it to something else? Sending bools as strings often causes confusion and I like to avoid it where possible. If you want me to alter it, do you have a preferred method for this? I would probably set it to `on` or `off`.

tollmanz commented on plugins/mobile.js:
```diff
@@ -4,22 +4,98 @@ Plugin to capture navigator.connection.type on browsers that support it
+
+			if (connection.addEventListener) {
+				connection.addEventListener("change", function() {
+					this.setConnectionInformation();
```
`this.setConnectionInformation()` throws an error in this context due to the `this` reference.

tollmanz replied:
fixed

bluesmoon replied:
I think `bind` is not compatible with older browsers.  Simplest solution is to just use `impl.setConnectionInformation` instead

tollmanz replied:
Yes of course. Fixed!

bluesmoon commented:
So just the one change needed.  Can you also verify that the API docs correctly reflect the new params?

tollmanz commented:
@bluesmoon Change has been made. I do not see any docs for the mobile plugin so I do not think updates need to be made.

I have two more questions:

- [ ] The `saveData` value is `true` or `false` according to the spec. Do you want that data sent as `(string) true/false` or should I alter it to something else? Sending bools as strings often causes confusion and I like to avoid it where possible. If you want me to alter it, do you have a preferred method for this? I would probably set it to on or off.
- [ ] [When setting the value](https://github.com/SOASTA/boomerang/pull/160/files#diff-453d0e3587548990865dc0c408ceda05R76), we do a truthy test like, `if (connection[k])`. Do you know specifically what you were guarding against here? Earlier in the code, we ensure that the prop exists. Do you think we need to do this again? This causes problems when the value is `0`. I've encountered this a few times with `rtt`. It will be set to `0` and thus `mob.rtt` will not get set.

bluesmoon commented:
Checking for the truthy value was specifically to prevent 0s, `null`s and `undefined`s. Instead if you'd like to do a `typeof` check, that would also be fine.

@nicjansma can you comment on the `saveData` question?

tollmanz commented:
> Checking for the truthy value was specifically to prevent 0s, nulls and undefineds. Instead if you'd like to do a typeof check, that would also be fine.

Thanks! I've updated it to allow strings, numbers, and booleans.

bluesmoon commented:
Looks good to me.
